File: solutions/js/13.js
/**
 * @param {string[]} factory - The factory layout
 * @returns {'completed'|'broken'|'loop'} Result of the gift journey
 */
function runFactory(factory) {
  let posX = 0;
  let posY = 0;
  const visitedPositions = new Set(`${posY}${posX}`);
  let actualMovement = factory[posY][posX];
  while (actualMovement !== undefined) {
    switch (actualMovement) {
      case ">":
        posX++;
        break;
      case "<":
        posX--;
        break;
      case "^":
        posY--;
        break;
      case "v":
        posY++;
        break;
    }
    actualMovement = factory[posY]?.[posX];
    if (actualMovement === ".") return "completed";
    const position = `${posY}${posX}`;
    if (visitedPositions.has(position)) return "loop";
    visitedPositions.add(position);
  }
  return "broken";
}

const factory1 = runFactory([">>."]); // 'completed'
console.log(factory1);

const factory2 = runFactory([">>>"]); // 'broken'
console.log(factory2);

const factory3 = runFactory([">><"]); // 'loop'
console.log(factory3);

const factory4 = runFactory([">>v", "..<"]); // 'completed'
console.log(factory4);

const factory5 = runFactory([">>v", "<<<"]); // 'broken'
console.log(factory5);

const factory6 = runFactory([">v.", "^.."]); // 'completed'
console.log(factory6);

const factory7 = runFactory(["v.", "^."]); // 'loop'
console.log(factory7);

// TODO: Validate
/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 10??? 
  ops: 1???
*/

File: solutions/python/13.py
def run_factory(factory: list[str]) -> str:
    pos_x = 0
    pos_y = 0
    visited_positions = {f"{pos_y}{pos_x}"}
    while ((0 <= pos_y < len(factory)) and (0 <= pos_x < len(factory[0]))):
        actual_movement = factory[pos_y][pos_x]
        if (actual_movement == ">"):
            pos_x += 1
        elif (actual_movement == "<"):
            pos_x -= 1
        elif (actual_movement == "^"):
            pos_y -= 1
        elif (actual_movement == "v"):
            pos_y += 1

        if (actual_movement == "."):
            return "completed"

        position = f"{pos_y}{pos_x}"
        if (position in visited_positions):
            return "loop"
        visited_positions.add(position)

    return "broken"


factory1 = run_factory([">>."])  # 'completed'
print(factory1)

factory2 = run_factory([">>>"])  # 'broken'
print(factory2)

factory3 = run_factory([">><"])  # 'loop'
print(factory3)

factory4 = run_factory([">>v", "..<"])  # 'completed'
print(factory4)

factory5 = run_factory([">>v", "<<<"])  # 'broken'
print(factory5)

factory6 = run_factory([">v.", "^.."])  # 'completed'
print(factory6)

factory7 = run_factory(["v.", "^."])  # 'loop'
print(factory7)

# TODO: Validate
"""
  Score: ★★★★★★★ (3 bonus stars for solving it today!)
  Points: 10??? 
  ops: 1???
"""
